tripleo-heat-templates-diff/apply_patch.py

```diff
@@ -40,44 +41,69 @@
 PRE_VERSION_3_DIFF = "diff_OSPD10_5.3.3-1"
 POST_VERSION_3_DIFF = "diff_OSPD10_5.3.10-17"
 
-if len(sys.argv) != 1:
-    print "Usage: %s" % sys.argv[0]
-    sys.exit(1)
 
 def version_compare((e1, v1, r1), (e2, v2, r2)):
     return rpm.labelCompare((e1, v1, r1), (e2, v2, r2))
 
-version = subprocess.check_output(['rpm', '-qa', 'openstack-tripleo-heat-templates'])
 
-(e0, v0, r0) = stringToVersion(version)
-(e1, v1, r1) = stringToVersion(VERSION_1_CHECK)
-(e2, v2, r2) = stringToVersion(VERSION_2_CHECK)
-(e3, v3, r3) = stringToVersion(VERSION_3_CHECK)
-(e4, v4, r4) = stringToVersion(VERSION_4_CHECK)
+def file_exists(filename):
+    if not os.path.exists(filename):
+        print "File: %s is not present in the current directory. " \
+              "Please check!!" % filename
+        sys.exit(1)
+    return filename
 
-args = "patch -p0 -N -d /usr/share"
 
-# Compare versions
-version_1_rc = version_compare((e0, v0, r0), (e1, v1, r1))
-version_2_rc = version_compare((e0, v0, r0), (e2, v2, r2))
-version_3_rc = version_compare((e0, v0, r0), (e3, v3, r3))
-version_4_rc = version_compare((e0, v0, r0), (e4, v4, r4))
-if version_1_rc < 0:
-    args = args + " < " + PRE_VERSION_1_DIFF
+def main():
 
-elif version_1_rc >= 0 and version_2_rc < 0:
-    args = args + " < " + PRE_VERSION_2_DIFF
+    if len(sys.argv) != 1:
+        print "Usage: %s" % sys.argv[0]
+        sys.exit(1)
 
-elif version_2_rc >= 0 and version_3_rc <= 0:
-    args = args + " < " + PRE_VERSION_3_DIFF
+    version = subprocess.check_output(
+        ['rpm', '-qa', 'openstack-tripleo-heat-templates']
+    )
 
-elif version_3_rc > 0 and version_4_rc <= 0:
-    args = args + " < " + POST_VERSION_3_DIFF
+    (e0, v0, r0) = stringToVersion(version)
+    (e1, v1, r1) = stringToVersion(VERSION_1_CHECK)
+    (e2, v2, r2) = stringToVersion(VERSION_2_CHECK)
+    (e3, v3, r3) = stringToVersion(VERSION_3_CHECK)
+    (e4, v4, r4) = stringToVersion(VERSION_4_CHECK)
 
-elif version_4_rc > 0:
-    print "Not supported for %s" % version
-    sys.exit(1)
+    args = "patch -p0 -N -d /usr/share"
 
-# Apply appropriate diff
-print "Applying: %s" % args
-subprocess.call(args, shell=True)
+    # Compare versions
+    version_1_rc = version_compare((e0, v0, r0), (e1, v1, r1))
+    version_2_rc = version_compare((e0, v0, r0), (e2, v2, r2))
+    version_3_rc = version_compare((e0, v0, r0), (e3, v3, r3))
+    version_4_rc = version_compare((e0, v0, r0), (e4, v4, r4))
+
+    if version_1_rc < 0:
+        args = args + " < " + file_exists(PRE_VERSION_1_DIFF)
+
+    elif version_1_rc >= 0 and version_2_rc < 0:
+        args = args + " < " + file_exists(PRE_VERSION_2_DIFF)
+
+    elif version_2_rc >= 0 and version_3_rc <= 0:
+        args = args + " < " + file_exists(PRE_VERSION_3_DIFF)
+
+    elif version_3_rc > 0 and version_4_rc <= 0:
+        args = args + " < " + file_exists(POST_VERSION_3_DIFF)
```

Comment:
you can write this a bit more generic i assume by setting up a dict of version_x_rc keys with diff file values and implement this using a loop going over the keys

Reply:
Can be done when I combine the document.